### ISSUE TITLE
Java: java/unsafe-deserialization doesn't detect use of javax.jms.ObjectMessage.getObject()

### ISSUE THREAD

**Issue by ebickle** [github/codeql#14569]
**Description of the issue**

The JMS type `javax.jms.ObjectMessage` performs deserialization using its `.getObject()` function. Historically there have been CVEs resulting from this functionality, particularly when JMS implementations do not provide any type restrictions: 
* [CVE-2015-5254](https://nvd.nist.gov/vuln/detail/CVE-2015-5254)
* [CVE-2016-4978](https://nvd.nist.gov/vuln/detail/CVE-2016-4978)
* https://github.com/rabbitmq/rabbitmq-jms-client/issues/3

The nature of these CVEs are very similar to the description of the CodeQL [java/unsafe-deserialization](https://codeql.github.com/codeql-query-help/java/java-unsafe-deserialization/) query - untrusted data flowing into a deserialization call. 

I wasn't able to get CodeQL to detect a vulnerability when remote input (e.g. `javax.jms.MessageListener.onMessage(Message message)`) flows into `javax.jms.ObjectMessage.getObject()`. The only reference to `ObjectMessage` that I see in the CodeQL source code is a `summaryModel` extension, but nothing in the `unsafeDeserialization` predicate of `java/ql/lib/semmle/java/security/UnsafeDeserializationQuery.qll`.

Can someone please confirm? It seems like a common vulnerability that the java/unsafe-deserialization query is missing.

Some very rough sample code:
```
import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.MessageListener;
import javax.jms.ObjectMessage;

public class Receiver implements MessageListener {
    public void onMessage(Message message) {
        try {
            ObjectMessage msg = (ObjectMessage)message;
            String text = msg.getObject().toString();
        } catch (JMSException exception) {
        }
    }
}
```

**Comment by atorralba** [comment]
Hi @ebickle, thanks for reporting this issue.

I've been looking a bit into this, and the problem is that it's not as straightforward as it may seem initially. Although is true that calling `ObjectMessage.getObject` on an untrusted `ObjectMessage` is dangerous, its actual impact depends on the JMS implementation. I ran some quick CodeQL queries at scale to validate this — for example:

* Amazon SQS JMS implementation is vulnerable, since it directly deserializes the object from a Base64 string: https://github.com/awslabs/amazon-sqs-java-messaging-lib/blob/4cb91355cb92d9361a2179233c9db89383b1299e/src/main/java/com/amazon/sqs/javamessaging/message/SQSObjectMessage.java#L98
* On the other hand, Apache ActiveMQ Artemis implements a system of allow/disallow lists to prevent this kind of attack, precisely after CVE-2016-4978 got reported and fixed: https://github.com/apache/activemq-artemis/blob/13a3e84e78bc955e8e7c812e35ee0ec9ead3e80f/artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQObjectMessage.java#L134

As I understand it, this means that, if we wrote a query for this code pattern, we wouldn't be able to tell with high precision whether the issue is exploitable or not, because it would depend on the project dependencies and the specific implementation of JMS used.

Now, we could err on the side of caution and write the query anyway, letting users dismiss/mark as FP those issues they are sure are using a safe implementation. But the problem is that we wouldn't be able to recommend any other fix: you either don't call `ObjectMessage.getObject` at all, or you change your implementation to be a safe one (which could already be the case). So I see how this could quickly turn into something confusing for users.

Now, I'm not saying this query is definitely a bad idea — on the contrary, I see the potential, but we need to be cautious with FPs and ineffective recommendations. I guess one quick thing we could do is not running this query by default (i.e. make it `medium` precision), and let users who want more results enable it explicitly if they want.

Happy to hear other opinions here (including yours @ebickle!).
